server2/Arcadia.Assistant/Arcadia.Assistant.Notifications/Notifications.cs

```diff
@@ -0,0 +1,188 @@
+                    Type = notificationMessage
+                        .NotificationTemplate // TODO: calculate correct value or remove this property
+                }
+            };
+
+            foreach (var deviceToken in deviceTokens.Values)
+            {
+                await this.pushNotifications.SendPushNotification(deviceToken.ToArray(), notificationContent,
+                    cancellationToken);
+            }
+        }
+
+        private async Task<(EmployeeId, UserPreferences)> GetUserPreferences(
+            EmployeeId id, CancellationToken cancellationToken)
+        {
+            var preferences = await this.userPreferences.ForEmployee(id).Get(cancellationToken);
+            return (id, preferences);
+        }
+
+        private bool IsPushNotification(
+            IDictionary<EmployeeId, UserPreferences> userPreferencesDictionary, EmployeeId employeeId)
+        {
+            if (userPreferencesDictionary.TryGetValue(employeeId, out var pref))
+            {
+                return pref.PushNotifications;
+            }
+
+            return false;
+        }
+
+        private async Task<IDictionary<EmployeeId, IReadOnlyCollection<DeviceRegistryEntry>>> GetDeviceTokens(
+            IReadOnlyCollection<EmployeeId> employeeIds, CancellationToken cancellationToken)
+        {
+            var employeePreferences = (await Task.WhenAll(employeeIds
+                    .Distinct()
+                    .Select(x => this.GetUserPreferences(x, cancellationToken))))
+                .ToDictionary(x => x.Item1, x => x.Item2);
+            var activeEmployees = employeeIds.Where(x => this.IsPushNotification(employeePreferences, x));
+            var tokens = await this.deviceRegistry.GetDeviceRegistryByEmployeeList(activeEmployees, cancellationToken);
+            return tokens.Keys.SelectMany(k => tokens[k].Select(x =>
+                    new
+                    {
+                        Key = k,
+                        Val = x
+                    }))
+                .GroupBy(x => x.Key)
+                .ToDictionary(x => x.Key,
+                    x => (IReadOnlyCollection<DeviceRegistryEntry>)x.Select(a => a.Val).ToList().AsReadOnly());
+        }
+
+        /// <summary>
+        ///     Optional override to create listeners (e.g., TCP, HTTP) for this service replica to handle client or user requests.
+        /// </summary>
+        /// <returns>A collection of listeners.</returns>
+        protected override IEnumerable<ServiceInstanceListener> CreateServiceInstanceListeners()
+        {
+            return this.CreateServiceRemotingInstanceListeners();
+        }
+
+        /// <summary>
+        ///     This is the main entry point for your service instance.
+        /// </summary>
+        /// <param name="cancellationToken">Canceled when Service Fabric needs to shut down this service instance.</param>
+        protected override async Task RunAsync(CancellationToken cancellationToken)
```

Comment:
what's the potential use of this loop? Maybe just remove it completely?

Reply:
Removed